design/proposals/modular-components.md

```diff
@@ -0,0 +1,234 @@
+   environment.
+
+## Proposal
+
+1. Retire `main.jsonnet` and `base.libsonnet`.
+   * The purpose of these files are often confusing to users. The goal of
+   `main.jsonnet` was to act as a way for environments to provide
+   environment-specific extensibility to components. However, this only works
+   if the component is written in jsonnet, and also requires the user to have
+   an advanced understanding of jsonnet.
+   * Furthermore, with the existing model, all components are operated on
+   during `apply`, `delete`, etc. If the components vary drastically between
+   environments, it is not convenient to "subtract" these components from each
+   environment's `main.jsonnet`.
+   * `base.libsonnet` also currently serves no purpose besides acting as an entry
+   point that _imports_ all components.
+
+2. Consolidate environment metadata files.
+   * Each environment's `.metadata` directory currently store 3 files related to
+   the generated `ksonnet-lib`. This is unecessary and also costly as the
+   number of environments grow. We needn't store multiple copies of the same
+   API version on disk. The Kubernetes API version that each environment uses
+   can be recorded in the environment specification file. The metadata files
+   can be locally cached once per k8s API version in `lib`.
+
+3. A component should not be uniquely defined by it's base file name.
+   * The existing implementation implies that there cannot be a `foo.jsonnet`
+   and a `foo.yaml`, since the base file name is shared.
+
+4. Handle environment specifications in `app.yaml` and retire `spec.json`.
+   * Multiple files and areas for configuration can cause confusions for users.
+   In order to keep configurations simple, I propose that all ksonnet-specific
+   _app-level_ configurations reside in the `app.yaml`.
+
+`app.yaml` would add an _environments_ section as follows:
+
+```
+apiVersion: 0.0.1
+kind: ksonnet.io/app
+name: config
+version: 0.0.1
+registries:
+  incubator:
+    gitVersion:
+      commitSha: ea3408d44c2d8ea4d321364e5533d5c60e74bce0
+      refSpec: master
+    protocol: github
+    uri: github.com/ksonnet/parts/tree/master/incubator
+environments:
+  dev:
+    k8sVersion: 1.9.0
+    server: https://my-dev-server.com
+    namespaces:
+      - dev-foo
+      - dev-bar
+    targets:
+      - auth
+      - dev
+  staging:
+    k8sVersion: 1.8.0
+    server: https://my-staging-server.com
+    namespaces:
+      - staging
+    targets:
+      - redis.jsonnet
+      - auth
+      - prod
+```
+
+The _environments_ section operates similarly to the existing `spec.json` with
+the additions of:
+
+1. Support for multiple environment namespaces. Deployments will happen across
+   all the namespaces for that environment. If a user wishes to have separate
+   deployments to separate namespaces, these deployment groups should be
+   modeled as two separate environments.
+2. The introduction of _targets_. _Targets_ are the component targets that an
```

Comment:
Is there a place where all components are available as a single objects (much like now retired `base.libsonnet`?). It was handy to have as CI pipeline could drop arbitrary code there and perform various validations and mandatory transformations (such as adding annotations and labels to ALL objects without need for modifying any components).

Reply:
Unrelated to the above comment, I think the concept of `targets` is hugely important. This is one area where I found ksonnet seemed to be lacking/difficult. It's very common for me in my use-cases (as described in your user stories above) to want to omit deploying specific resources and components typically based on environment.

I do agree with redbaron that being able to basically transform all resources in some way would be nice, I can see that being useful to set ownerRefs/annotations across all objects if an operator was running ksonnet or something.

Reply:
We don't necessarily need to retire `base.libsonnet`, especially if a strong use case like this one presents. Targets can operate in parallel.